design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,228 @@
+each vendor provides 100 or more BIOS Settings, its not practical to store
+these in the BareMetalHost CRD. The new CRD will be named
+`HostFirmwareSettings` and will consist of the following:
+
+- `currentFirmwareSettings` - the current BIOS Settings retrieved from Ironic
+  via the bios API will be stored in the Status section. This dataset includes
+  BIOS Registry information which provides a schema for the settings.
+- `writeableFirmwareSettings` - BIOS Settings to write to the host via Ironic
+  will be stored in the Spec section. If it is empty, it will be initially
+  populated with the name/value pairs from `currentFirmwareSettings` that do
+  not have the `read-only` or `unique` flag set.
+
+The BIOS Settings are retrieved from the BMC by Ironic and cached
+whenever the node moves to `manageable` or `cleaning`, or when the settings
+are updated. Recent changes to Ironic (see the Ironic proposal
+[BIOS Registry](https://review.opendev.org/c/openstack/ironic-specs/+/774681)
+added an additional query of the BIOS Registry.
+
+The baremetal-operator can get the BIOS Settings from Ironic after the node
+has been moved to manageable via the `v1/nodes/{node_ident}/bios?detail=True`
+API.  This API provides a list of BIOS Settings (aka `Attributes`) in JSON and
+each setting includes the following data:
+
+- Name
+- Value
+- Registry information
+  - Type (e.g. enum, string, integer, boolean, or password)
+  - ReadOnly (a boolean indicating whether the attribute is read only)
+  - Unique (indicating the value is specific to the host where it was
+    retrieved)
+  - Allowable values for enum type attributes
+  - Minimum and maximum values (for numeric attributes)
+  - Minimum and maximum character lengths (for string attributes)
+
+Regarding the baremetal-operator states for when the data will be updated:
+
+- The node first transitions to manageable during the `Registration` state, so
+  at the end of that state `currentFirmwareSettings` will be populated.
+- Settings can be updated during the `Preparing` state, so at the end of that
+  state the settings will also be retrieved and used to update
+  `currentFirmwareSettings`.
+
+Upon retrieving the BIOS Settigs from Ironic, the baremetal-operator will
+copy the name/value pairs to the `writeableFirmwareSettings` field in
+`HostFirwareSettingsSpec` if that field is empty. Settings that have the
+`read-only` flag set will not be copied as these settings cannot be written
+to the BMC. Likewise, settings that have the `unique` flag set will also not
+be copied, these are settings like serial numbers or product ids for example,
+that should not be written to other servers.
+
+A user can update `writeableFirmwareSettings` to the desired values for the
+settings.  Likewise, `writeableFirmwareSettings` can be copied to multiple hosts
+of the same type to effectively clone the settings from a reference host and
+ensuring that all similar systems have the same BIOS configuration,
+i.e. "bulk set".
```

Comment:
Can you please elaborate a little bit more how the bulk copy process will work? From the description so far I wasn't able to find any templating feature so far, as mentioned at the beginning of the proposal

Reply:
There is one CRD instance per host, the same as BMH and in the same namespace. Each CRD instance will hold multiple BIOS settings, from a survey of common vendors this seems to be 100-200 settings.

As for the bulk copy, the plan was to just use the kubectl (oc) edit and apply commands to update the settings on a host from a known good host. I'm not aware if there is a simpler way to do this. I'l add that to the spec

Reply:
I don't want to drag away from scope this proposal, but I really see a potential opportunity here, so throwing in the idea just for the sake of the discussion. In order to really push for an (automated) bulk apply of the settings, the CRD could be enhanced by adding new selector fields for:

* The source - the host from where the settings will be copied
* The destination hosts - the targets where the settings will be copied to

(Selector fields could be labels or hints). In this case a dedicated controller would monitor the namespace for new CRD tagged with the _source_ and, in case, it will generate new CRD for the selected targets, by copying the allowed settings (at this point the BMO will operate as described in the proposal, going on with the communication to Ironic).
The controller could also take care of additional checks, for example to generate the CRD only for the select targets that also matches the hw vendors.
In this way we could have a quick way to effectively apply in bulk the settings without a tedious/error prone manual operation - especially for a large number of hosts.

Reply:
Andrea - I think this is a great idea and, as it turns out, a feature some customers have been asking for per discussion with our PMs.

To summarize our offline conversation regarding this:
- this current proposal is really for a minimal BIOS set feature that requires manual configuration for bulk set of the spec fields to other systems of the same type
- we should indicate in this proposal that there is a future feature to automate BIOS setting using a new operator that would take as input a list of systems of the same type and attempt to keep the settings in sync. This feature will require a new doc and potentially some new fields to the CRD although it will not be a breaking change to the existing CRD as it will not add mandatory fields
 
I'll update the doc to reflect this discussion.

Reply:
I'm not sure about the mechanics here...

> the CRD could be enhanced by adding new selector fields for:
> 
>     * The source - the host from where the settings will be copied

I don't think this works as a selector - if there's more than one matching host, what do you choose? I could see it working as just a straight link, although it would be inefficient to implement.

>     * The destination hosts - the targets where the settings will be copied to

Why both? Doing it in this direction feels a bit wrong - another Host can opt you in to receiving settings changes. There may even be some RBAC implications, although I'm not sure.

> In this way we could have a quick way to effectively apply in bulk the settings without a tedious/error prone manual operation - especially for a large number of hosts.

That would certainly be a nice improvement for the future.